src/puppetlabs/puppetdb/threadpool.clj

```diff
@@ -0,0 +1,119 @@
+  [pool-thread-factory]
+  (-> (BasicThreadFactory$Builder.)
+      (.wrappedFactory pool-thread-factory)
+      (.namingPattern "cmd-proc-thread-%d")
+      (.uncaughtExceptionHandler logging-exception-handler)
+      (.daemon true)
+      .build))
+
+(defrecord GatedThreadpool [^Semaphore semaphore ^ExecutorService threadpool])
+
+(defn create-threadpool
+  "Creates an unbounded threadpool with the intent that access to the
+  threadpool is bounded by the semaphore. Implicitly the threadpool is
+  bounded by `size`, but since the semaphore is handling that aspect,
+  it's more efficient to use an unbounded pool and not duplicate the
+  constraint in both the semaphore and the threadpool"
+  [size]
+  (let [threadpool (ThreadPoolExecutor. 1
+                                        Integer/MAX_VALUE
+                                        1
+                                        TimeUnit/MINUTES
+                                        (SynchronousQueue.))]
+    (->> threadpool
+         .getThreadFactory
+         thread-factory
+         (.setThreadFactory threadpool))
+
+    (map->GatedThreadpool
+     {:semaphore (Semaphore. size)
+      :threadpool threadpool})))
+
+(defn shutdown
+  "Shuts down the gated threadpool, ensuring in-flight work is
+  complete before tearing it down. Currently waits up to 5 minutes for
+  that work to complete"
+  [{:keys [^Semaphore semaphore ^ExecutorService threadpool]}]
+
+  ;; Prevent new work from being accepted
+  (.drainPermits semaphore)
+
+  ;; Shutdown the threadpool, allowing in-flight work to finish
+  (.shutdown threadpool)
+
+  ;; This will block, waiting for the threadpool to shutdown
+  (when-not (.awaitTermination threadpool 5 java.util.concurrent.TimeUnit/MINUTES)
+
+    (log/warn "Attempted to shutdown threadpool, not stopped after 5 minutes, forcibly shutting it down")
+
+    ;; This will force the shutdown of the threadpool and will
+    ;; not allow current threads to finish
+    (.shutdownNow threadpool)
+    (log/warn "Threadpool forcibly shutdown")))
+
+
+(defn call-on-threadpool
+  "Executes `f` on a gated threadpool. Calls `on-complete` with the
+  results of inoking `f`"
+  ([threadpool f]
+   (call-on-threadpool threadpool identity f))
+  ([{:keys [^Semaphore semaphore ^ExecutorService threadpool]}
+    on-complete f]
+   (.acquire semaphore)
+   (try
+     (.execute threadpool (fn []
+                            (try
+                              (on-complete (f))
+                              (finally
+                                (.release semaphore)))))
+     (catch RejectedExecutionException e
+       (log/error e "Message not submitted to command processing threadpool")
+       (.release semaphore)))))
+
+(defn exec-from-channel
+  "Executes `on-input` for each input found on `in-chan`, with a
+  threadpool of size `parallelism`. The results of the `on-input`
+  invocations will be put to the `out-chan` in the order it has
+  completed"
```

Comment:
This is the blocking version, right? Mention of that in docstring would be nice, as well as something about how it'll close out-chan for you.